pkg/prom/diagnostics.go

```diff
@@ -0,0 +1,68 @@
+package prom
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/kubecost/cost-model/pkg/env"
+	"github.com/kubecost/cost-model/pkg/log"
+	prometheus "github.com/prometheus/client_golang/api"
+)
+
+// QueuedPromRequest is a representation of a request waiting to be sent by the prometheus
+// client.
+type QueuedPromRequest struct {
+	Context   string `json:"context"`
+	Query     string `json:"query"`
+	QueueTime int64  `json:"queueTime"`
+}
+
+// PrometheusQueueState contains diagnostic information concerning the state of the prometheus request
+// queue
+type PrometheusQueueState struct {
+	QueuedRequests      []*QueuedPromRequest `json:"queuedRequests"`
+	OutboundRequests    int                  `json:"outboundRequests"`
+	TotalRequests       int                  `json:"totalRequests"`
+	MaxQueryConcurrency int                  `json:"maxQueryConcurrency"`
+}
+
+// GetPrometheusQueueState is a diagnostic function that probes the prometheus request queue and gathers
+// query, context, and queue statistics.
+func GetPrometheusQueueState(client prometheus.Client) (*PrometheusQueueState, error) {
+	rlpc, ok := client.(*RateLimitedPrometheusClient)
+	if !ok {
+		return nil, fmt.Errorf("Failed to get prometheus queue state for the provided client. Must be of type RateLimitedPrometheusClient.")
+	}
+
+	outbound := rlpc.TotalOutboundRequests()
+
+	requests := []*QueuedPromRequest{}
+	rlpc.queue.Each(func(_ int, entry interface{}) {
+		if req, ok := entry.(*workRequest); ok {
+			requests = append(requests, &QueuedPromRequest{
+				Context:   req.contextName,
+				Query:     req.query,
+				QueueTime: time.Since(req.start).Milliseconds(),
+			})
+		}
+	})
+
+	return &PrometheusQueueState{
+		QueuedRequests:      requests,
+		OutboundRequests:    outbound,
+		TotalRequests:       outbound + len(requests),
+		MaxQueryConcurrency: env.GetMaxQueryConcurrency(),
+	}, nil
```

Comment:
Should we lock the rlpc queue for the duration of this run? (I think that'd mess up `.Each`, but maybe a read-lock?)

Reply:
No, we don't need another lock.

Reply:
To provide further detail here, there is a very small chance that `TotalOutboundRequests()` may be off by `1` if the queue happens to process the next queue entry before `.Each`, The outbound requests are counted using an atomic increment so reads/writes are safe. To be honest, the main focus of this status endpoint is to display the contents of the queue and the counters are meant to be sort of a general summary of the contents. I think it's perfectly ok here to avoid more locking. `.Each` already disallows access to the queue.